File: ml_test_vectors/stateless/sigmoid/implementation.py
from typing import Optional

import numpy as np

from ml_test_vectors.types import Tensor


def forward(inp: Tensor) -> Tensor:
    x = np.array(inp, dtype="float32")
    return 1 / (1 + np.exp(-x, dtype="float32"))


def backward(inp: Tensor, output: Optional[Tensor]) -> Tensor:
    outp = forward(inp) if output is None else output
    x = np.array(outp, dtype="float32")
    return x * (1 - x)

File: ml_test_vectors/stateless/sigmoid/implementation.ts
import type {
	StatelessDerivativeFunction,
	StatelessFunction,
	Tensor,
} from "../../core";
import { isScalar } from "../../utils";

export const forward: StatelessFunction<Tensor, Tensor> = (input) =>
	isScalar(input) ? 1 / (1 + Math.exp(-input)) : input.map(forward);

export const backward: StatelessDerivativeFunction = (
	input,
	output = forward(input),
) => {
	if (isScalar(output)) {
		return output * (1 - output);
	}
	if (isScalar(input)) {
		throw new Error("Mismatched input/output");
	}
	return output.map((o, i) => backward(input[i], o));
};
